activesupport/lib/active_support/cache/coder.rb

```diff
@@ -88,6 +81,36 @@ def load(dumped)
 
         MARSHAL_SIGNATURE = "\x04\x08".b.freeze
 
+        class LazyEntry < Cache::Entry
+          def initialize(serializer, compressor, payload, type, **options)
+            super(payload, **options)
+            @serializer = serializer
+            @compressor = compressor
+            @type = type
+          end
+
+          def value
+            if @type
+              @value = @compressor.inflate(@value) if @compressor
+
+              if string_encoding = STRING_ENCODINGS[@type & ~COMPRESSED_FLAG]
+                @value.force_encoding(string_encoding)
+              else
```

Comment:
Ok, so I think if we extracted this to be a serializer, it would be perfect.

So basically if we had something like:

```ruby
module Utf8Serializer
  extend self
  def load(payload)
    payload.force_encoding(Encoding::UTF_8)
  end
end
STRING_SERIALIZERS = {
  0x02 => Utf8Serializer,
}  
```

This way you can check the type before calling `LazyEntry.new` and always pass a serializer, and don't need to pass the type.

Reply:
I've pushed two more commits (031a4e9d0d9070728d3ee80804ab4553afd41598 and b819095cf5fa1a30c05757344f79c0379c03dfe4).  The first implements `Utf8Serializer`, etc.  That seemed too cumbersome to me, so the 2nd commit refactors them as `StringDeserializer` instances.

It's worth pointing out that their `load` method mutates its argument, whereas typical `load` methods do not.  That does not affect our implementation, but it is a leaky abstraction.

Reply:
> so the 2nd commit refactors them as StringDeserializer instances.

Yes, good call, it's quite nice.

> It's worth pointing out that their load method mutates its argument,

That's the case too with the existing `Coder`. it's indeed a leaky abstraction, but here we know we the caller won't use the string, so it's a nice little perf gain.